airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py

```diff
@@ -111,19 +111,34 @@ def next(self) -> Optional[Mapping[str, Any]]:
         """
 
         try:
-            if self._current_slice is None:
-                self._current_slice = self._get_next_slice()
-                state_for_slice = self._cursor.select_state(self._current_slice)
-                if state_for_slice == {"__ab_full_refresh_sync_complete": True}:
-                    return None
-                else:
-                    return self._current_slice
             if self._read_state_from_cursor:
-                state_for_slice = self._cursor.select_state(self._current_slice)
-                if state_for_slice == {"__ab_full_refresh_sync_complete": True}:
-                    self._current_slice = self._get_next_slice()
+                if self._current_slice is None:
+                    next_slice = self._get_next_slice()
+                    state_for_slice = self._cursor.select_state(self._current_slice)
+                    if state_for_slice == {"__ab_full_refresh_sync_complete": True}:
+                        next_candidate_slice = None
+                        has_more = True
+                        while has_more:
+                            next_candidate_slice = self._get_next_slice()
+                            state_for_slice = self._cursor.select_state(next_candidate_slice)
+                            has_more = state_for_slice == {"__ab_full_refresh_sync_complete": True}
+                        self._current_slice = StreamSlice(cursor_slice=state_for_slice or {}, partition=next_candidate_slice.partition)
+                    else:
+                        self._current_slice = StreamSlice(cursor_slice=state_for_slice or {}, partition=next_slice.partition)
                 else:
-                    self._current_slice = StreamSlice(cursor_slice=state_for_slice or {}, partition=self._current_slice.partition)
+                    state_for_slice = self._cursor.select_state(self._current_slice)
+                    if state_for_slice == {"__ab_full_refresh_sync_complete": True}:
+                        # Skip every slice that already has the terminal complete value indicating that a previous attempt
+                        # successfully synced the slice
+                        next_candidate_slice = None
+                        has_more = True
+                        while has_more:
+                            next_candidate_slice = self._get_next_slice()
+                            state_for_slice = self._cursor.select_state(next_candidate_slice)
+                            has_more = state_for_slice == {"__ab_full_refresh_sync_complete": True}
+                        self._current_slice = StreamSlice(cursor_slice=state_for_slice or {}, partition=next_candidate_slice.partition)
```

Comment:
Can we have the dedicated `setter` and assign the value to the instance variable: `self._current_slice`, like:
```python
def set_current_slice(self, value) -> None:
   self._current_slice = value
```
Then reuse this within the code parts.

This clears up the code and makes it easier to follow.

Reply:
yeah i can actually just transition this into python getter/setter property pattern and replace, not a problem